### PR TITLE
[RPC] Reworked the gettransaction RPC method.

### PR DESCRIPTION
**Pull request by bokobza**
Made the implementation look more like bitcoin core's.

Closes https://github.com/stratisproject/StratisBitcoinFullNode/issues/2978
Closes https://stratisplatformuk.visualstudio.com/StratisBitcoinFullNode/_workitems/edit/2971
Related to https://github.com/stratisproject/StratisBitcoinFullNode/issues/2985
Related to https://github.com/stratisproject/StratisBitcoinFullNode/issues/1761
Related to https://github.com/stratisproject/StratisBitcoinFullNode/issues/2758

File used to test parity with StratisX which is close enough to bitcoin Core to be useful.
[GetTransaction.txt](https://github.com/stratisproject/StratisBitcoinFullNode/files/2846035/GetTransaction.txt)

Also contains the fix included in https://github.com/stratisproject/StratisBitcoinFullNode/pull/3211

cc: @mikedennis @sondreb @zeptin

### CODE REVIEW

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+using Stratis.Bitcoin.Features.Wallet.Models;
+using Stratis.Bitcoin.IntegrationTests.Common;
+using Stratis.Bitcoin.IntegrationTests.Common.EnvironmentMockUpHelpers;
+using Stratis.Bitcoin.IntegrationTests.Common.ReadyData;
+using Stratis.Bitcoin.Networks;
+using Xunit;
+
+namespace Stratis.Bitcoin.IntegrationTests.Wallet
+{
+    public class WalletRPCControllerTest
```
Comment (mikedennis) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Very nice set of tests... wondering whether we should be moving some of the tests from Stratis.Bitcoin.IntegrationTests.RPC related to the WalletRPCController to this class? I could do that in a separate PR later if you think it's a good idea. Does WithBlockChainReadyData() avoid the need to use "generate" to make some transactions?

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+                // Arrange.
+                CoreNode node = builder.CreateStratisPosNode(this.network).WithReadyBlockchainData(ReadyBlockchain.StratisRegTest150Miner).Start();
+
+                RPCClient rpc = node.CreateRPCClient();
+
+                try
+                {
+                    await rpc.SendCommandAsync(RPCOperations.gettransaction, txId);
+                }
+                catch (RPCException exception)
```
[the remark this answers is not in the document]
Comment (bokobza) [reply src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: ah yes, thanks, I'll change that.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs (hunk 1 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Yeah these tests should cover everything really.
Well, this is RPC and wallet so it's hard to know whether they should go with RPC or wallet.
I'm gonna go through all these methods now to triple check they're ok and if I write some tests I'll drop them in there.

With WithBlockChainReadyData, what we did is that we created a blockchain with 10/100/150 blocks already mined, so that we don't have to mine every time we need some data. We've gained something like 45 minutes on the build already.

File: src/Stratis.Bitcoin.Features.Wallet/Models/GetTransactionModel.cs [hunk 3]
```diff
@@ -1,41 +1,156 @@
...
-
-        [JsonProperty("amount")]
-        public Money Amount { get; set; }
-    }
-
     /// <summary>Model for RPC method gettransaction.</summary>
     public class GetTransactionModel
     {
+        /// <summary>
+        /// The total amount recieve or spent in this transaction.
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/Models/GetTransactionModel.cs]: "received"

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+                RPCClient rpcReceivingNode = receivingNode.CreateRPCClient();
+                RPCResponse txReceivingWallet = rpcReceivingNode.SendCommand(RPCOperations.gettransaction, txId.ToString());
+
+                RPCClient rpcSendingNode = sendingNode.CreateRPCClient();
+                RPCResponse txSendingWallet = rpcSendingNode.SendCommand(RPCOperations.gettransaction, txId.ToString());
+
+                // Assert.
+                GetTransactionModel resultSendingWallet = txSendingWallet.Result.ToObject<GetTransactionModel>();
+                resultSendingWallet.Amount.Should().Be(new Money(-100000000));
+                //resultSendingWallet.Fee.Should().Be(new Money(100000000));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Remove commented out code.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+                resultSendingWallet.BlockIndex.Should().BeNull();
+                resultSendingWallet.BlockTime.Should().BeNull();
+                resultSendingWallet.TimeReceived.Should().BeAfter(DateTimeOffset.Now - TimeSpan.FromMinutes(1));
+                resultSendingWallet.TransactionTime.Should().Be(Utils.UnixTimeToDateTime(trx.Time));
+                resultSendingWallet.Details.Count.Should().Be(1);
+
+                GetTransactionDetailsModel detailsSendingWallet = resultSendingWallet.Details.Single();
+                detailsSendingWallet.Address.Should().Be(unusedaddresses.Single());
+                detailsSendingWallet.Amount.Should().Be(new Money(-100000000));
+                //detailsSendingWallet.Fee.Should().Be(new Money(100000000));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Remove commented out code.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 6]
```diff
@@ -0,0 +1,686 @@
...
+                RPCClient rpcReceivingNode = receivingNode.CreateRPCClient();
+                RPCResponse txReceivingWallet = rpcReceivingNode.SendCommand(RPCOperations.gettransaction, txId.ToString());
+
+                RPCClient rpcSendingNode = sendingNode.CreateRPCClient();
+                RPCResponse txSendingWallet = rpcSendingNode.SendCommand(RPCOperations.gettransaction, txId.ToString());
+
+                // Assert.
+                GetTransactionModel resultSendingWallet = txSendingWallet.Result.ToObject<GetTransactionModel>();
+                resultSendingWallet.Amount.Should().Be(new Money(-100000000));
+                //resultSendingWallet.Fee.Should().Be(new Money(100000000));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Remove commented out code.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 7]
```diff
@@ -0,0 +1,686 @@
...
+                RPCClient rpcReceivingNode = receivingNode.CreateRPCClient();
+                RPCResponse txReceivingWallet = rpcReceivingNode.SendCommand(RPCOperations.gettransaction, txId.ToString());
+
+                RPCClient rpcSendingNode = sendingNode.CreateRPCClient();
+                RPCResponse txSendingWallet = rpcSendingNode.SendCommand(RPCOperations.gettransaction, txId.ToString());
+
+                // Assert.
+                GetTransactionModel resultSendingWallet = txSendingWallet.Result.ToObject<GetTransactionModel>();
+                resultSendingWallet.Amount.Should().Be(new Money(-200000000));
+                //resultSendingWallet.Fee.Should().Be(new Money(-100000000));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Remove commented out code.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 8]
```diff
@@ -0,0 +1,686 @@
...
+
+                GetTransactionDetailsModel detailsReceivingWallet = resultSendingWallet.Details.Single(d => d.Category == GetTransactionDetailsCategoryModel.Receive);
+                detailsReceivingWallet.Address.Should().Be(unusedaddresses.Single());
+                detailsReceivingWallet.Amount.Should().Be(new Money(100000000));
+                detailsReceivingWallet.Fee.Should().BeNull();
+
+                GetTransactionDetailsModel secondDetailsReceivingWallet = resultSendingWallet.Details.Single(d => d.Category == GetTransactionDetailsCategoryModel.Send);
+                secondDetailsReceivingWallet.Address.Should().Be(unusedaddresses.Single());
+                secondDetailsReceivingWallet.Amount.Should().Be(new Money(-100000000));
+                //secondDetailsReceivingWallet.Fee.Should().Be(new Money(-100000000));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Remove commented out code.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs [hunk 9]
```diff
@@ -0,0 +1,686 @@
...
+
+                Transaction trx = this.network.Consensus.ConsensusFactory.CreateTransaction(buildTransactionModel.Hex);
+
+                RPCClient rpcSendingNode = sendingNode.CreateRPCClient();
+                RPCResponse txSendingWallet = rpcSendingNode.SendCommand(RPCOperations.gettransaction, txId.ToString());
+
+                // Assert.
+                GetTransactionModel resultSendingWallet = txSendingWallet.Result.ToObject<GetTransactionModel>();
+                resultSendingWallet.Amount.Should().Be(new Money(-9800000200000000));
+                //resultSendingWallet.Fee.Should().Be(new Money(-100000000));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletRPCControllerTest.cs]: Remove commented out code.

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 10]
```diff
@@ -176,52 +180,140 @@ public decimal GetBalance(string accountName, int minConfirmations = 0)
...
+        /// Uses the default wallet if specified, or the first wallet found.
         /// </summary>
-        /// <param name="txid">Transaction identifier to find.</param>
+        /// <param name="txid">Identifier of the transaction to find.</param>
         /// <returns>Transaction information.</returns>
         [ActionName("gettransaction")]
-        [ActionDescription("Gets a transaction from the wallet.")]
-        public GetTransactionModel GetTransaction(string txid)
+        [ActionDescription("Get detailed information about an in-wallet transaction.")]
+        public async Task<GetTransactionModel> GetTransactionAsync(string txid)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: So for me personally I would split this method up a bit as it is very long and hard to follow. I just prefer smaller methods and it is easier compartmentalize the logic. It might personal preference so its not a blocker but what do you think?

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: The Fee is something to which we don't have access just yet, but I'm keen to keep it there so that we can review when it's available.

The method is long but it does one thing. Splitting it in many won't help much I'm afraid. I think it even might make it harder to follow.

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: OK cool, if this is the case maybe just add `// TODO: Fee needs to be implemented....`

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: I have it on here
image

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: I'll add a more meaningful meesage.

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: OK cool, then just add TODO to the tests, so that we know that they also need to be implemented.

**Comment by sondreb** [comment]
I think this looks nice, I think on this method the primary change I did was include the .Confirmations on the result, which you have done now.

I have implemented "getwalletinfo", which is still missing and I can make a PR with my implementation for discussions.

Here is the specification for getwalletinfo: https://bitcoin.org/en/developer-reference#getwalletinfo

I have implemented the values:

Balance, WalletName, WalletVersion, UnConfirmedBalance, ImmatureBalance

**Comment by bokobza** [comment]
@sondreb please submit what you have and we can discuss.
With RPc, we're trying to be as close as possible to what Bitcoin Core is doing as most users already have a BTC implementation in place.

@fassadlr I've restarted the MacOS build.

**Comment by sondreb** [comment]
@bokobza I will submit my "getwalletinfo" implementation in a few days, but please include the "account" field on details type (GetTransactionDetailsModel) as that is required and it can simply be empty string.
